source/structuring.rst

```diff
@@ -14,402 +14,545 @@ functions is better than copying-and-pasting duplicate bits of code
 
-``$ref`` elements may be used to create recursive schemas that refer to themselves.
-For example, you might have a ``person`` schema that has an array of ``children``, each of which are also ``person`` instances.
+1) find the value of the key ``properties``
+2) within that object, find the value of the key ``street_address``
+
+The URI
+``https://example.com/schemas/address#/properties/street_address``
+identifies the highlighted subschema in the following schema.
 
 .. schema_example::
 
     {
-      "$schema": "http://json-schema.org/draft-07/schema#",
-
-      "definitions": {
-        "person": {
-          "type": "object",
-          "properties": {
-            "name": { "type": "string" },
-            "children": {
-              "type": "array",
-    *          "items": { "$ref": "#/definitions/person" },
-              "default": []
-            }
-          }
-        }
-      },
+      "$id": "https://example.com/schemas/address",
 
       "type": "object",
-
       "properties": {
-        "person": { "$ref": "#/definitions/person" }
-      }
-    }
-    --
-    // A snippet of the British royal family tree
-    {
-      "person": {
-        "name": "Elizabeth",
-        "children": [
-          {
-            "name": "Charles",
-            "children": [
-              {
-                "name": "William",
-                "children": [
-                  { "name": "George" },
-                  { "name": "Charlotte" }
-                ]
-              },
-              {
-                "name": "Harry"
-              }
-            ]
-          }
-        ]
-      }
+        "street_address":
+      *    { "type": "string" },
+        "city": { "type": "string" },
+        "state": { "type": "string" }
+      },
+      "required": ["street_address", "city", "state"]
     }
 
-Above, we created a schema that refers to another part of itself, effectively
-creating a "loop" in the validator, which is both allowed and useful. Note,
-however, that a loop of ``$ref`` schemas referring to one another could cause an
-infinite loop in the resolver, and is explicitly disallowed.
+.. index::
+   single: $id
+   single: named anchors
+   single: structuring; subschema identification; $id
+
+.. _anchor:
+
+Named Anchors
+~~~~~~~~~~~~~
+
+Another way to identify a subschema is to create a named anchor in
+the schema using the ``$id`` keyword and using that name in the URI
+fragment instead of a JSON Pointer.
+
+.. draft_specific::
+
+    --Draft 4
+    In Draft 4, ``$id`` is just ``id`` (without the dollar sign).
+
+When the ``$id`` keyword contains a URI fragment, the fragment defines
+a named anchor using the value of the fragment. Named anchors must
+start with a letter followed by any number of letters, digits, ``-``,
+``_``, ``:``, or ``.``.
+
+.. note::
+   If a named anchor is defined that doesn't follow these naming
+   rules, then behavior is undefined. Your anchors might work in some
+   implementation, but not others.
```

Comment:
This is precisely what draft-07 says about this case. I don't recall what later drafts say exactly, but if they call it an error rather than undefined, this will be updated when I get to documenting those drafts.

> The effect of using a fragment in "$id" that isn't blank or doesn't follow the plain name syntax is undefined.

Reply:
Well huh. OK then.